harness/tests/tensorboard/test_base.py

```diff
@@ -90,3 +94,53 @@ def test_illegal_type() -> None:
         tensorboard.build(
             env.det_cluster_id, env.det_experiment_id, env.det_trial_id, checkpoint_config
         )
+
+
+def test_upload_thread_normal_case() -> None:
```

Comment:
I'd recommend omitting these tests, as I don't think they are offering useful test coverage.

The first is mostly testing that queues work, and the second is mostly testing that try/finally works.

Reply:
Thank you for the comment Ryan! Let me share my thinking in adding these tests in more depth for discussion. My main goal of including tests are two folds:

## (1) to guard the expected behavior of the code from accidental changes in the future
`test_upload_thread_normal_case` test that the thread is able to dequeue and it would exit when a sentinel value is passed (which is one of the two main design decisions listed)

`test_upload_thread_exception_case` guards that the thread class would be able to handle exception thrown from the upload and not terminate, which is also one of the design decisions listed. If the thread would terminate from exception, then it is a problem as no more files will be synced from that point onwards. 

Both of these tests guard expected behaviors of the thread class as opposed to implementation details.

While adding comments to the code to explain and caution people is an option, I personally found that have tests guarding against those changes tend to be more effective.

## (2) To make sure a self-contained code piece works as expected
I tend to break down a larger task to self-contained smaller PRs. End to end test is often not possible for these smaller pieces and thus I use unit test to ensure the self-contained piece actually works before I go onto implementing the next piece.

What do you think?

Reply:
I agree with the principles you are citing, but I also think that the code you have is simple enough that it's really not going to be a problem.

It's an interesting discussion point, and definitely a subjective point too.  Does a functionality test that is smaller than an expected user behavior help?  What keeps a future developer from reading the code and saying, for example, "who cares if None is the sentinel", then they change the code and the test?  Clearly my opinion is subjective here.

I like focusing on code simplicity and following standard, obvious patterns, because I think that goes further for protecting code from future complexity (and bugs).  And I think this code is sufficiently simple and obvious to meet that need.  Another subjective opinion.

So, if you would like to keep the tests, as I do agree with your motivations, and as I generally like good test coverage, I will not ask you to remove the tests if you think they are worth it.  As long as the code keeps working in the future, that's really what's important.

Reply:
Many good points here Ryan, thanks for sharing your thoughts! I am sure we will have some good discussions on this topic in the future, more in depth than this comment thread can afford.

- I agree that code simplicity and following standard, obvious patterns are the most important elements to protecting code from future complexity (and bugs).
- "Does a functionality test that is smaller than an expected user behavior help?" is a good discussion point, I have yet to contemplate about this enough to have a very well informed opinion.
- "for example, "who cares if None is the sentinel", then they change the code and the test?" I would say this is actually desirable, as the test is making sure the thread can terminate properly, as long as the change preserve this behavior, it is totally ok, regardless of what sentinel value they choose. 

For now, I will keep these tests here as I think they are should work in the future and they guard some desired behaviors of the class. Also, it is also subjective as they give me confidence that my code is working as intended.